### PR TITLE
Move GlobalTracer from contrib to opentracing-java.

### PR DESCRIPTION
**Pull request by sjoerdtalsma**
There seems to be general consensus on [gitter](https://gitter.im/opentracing/public) to move the Java `GlobalTracer` into the `opentracing-java` project.

I took the liberty to use the version in https://github.com/opentracing-contrib/java-globaltracer/pull/8 for this PR as well. I propose to keep the discussion about that change confined to that PR as much as possible. I will make sure to keep them functionally aligned.

Discussion we could have here is:
- Should this be a separate module?
- Or should the `GlobalTracer` class be moved into the API module itself for even more exposure? 

Given the tendency to move things like `no-op` out of the API, my initial choice was a separate module with the `io.opentracing:opentracing-globaltracer` groupId:artifact identity.
This allows migration of the current 'contrib' variant by publishing a new _deprecated + delegated_ variant.

### CODE REVIEW

**Comment by bhs** [comment]
@sjoerdtalsma thank you for kicking this off!

IMO the separate module is the right approach, though I think it's fair game to have the OT-Java docs point directly towards it. I would also propose that we choose a Java package name that can stick if we want to move this into the core module in the future.

File: opentracing-globaltracer/src/main/java/io/opentracing/GlobalTracer.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+ * <ol type="a">
+ * <li>The last {@link #register(Tracer) registered} or {@link #update(UpdateFunction) updated} tracer
+ * always takes precedence.</li>
+ * <li>If no tracer was registered, one is looked up from the {@link ServiceLoader}.<br>
+ * The {@linkplain GlobalTracer} will not attempt to choose between implementations:</li>
+ * <li>If no single tracer service is found, the {@link io.opentracing.NoopTracer NoopTracer} will be used.</li>
+ * </ol>
+ */
+public final class GlobalTracer implements Tracer {
+    private static final Logger LOGGER = Logger.getLogger(GlobalTracer.class.getName());
```
Comment (frenchfrywpepper) [inline opentracing-globaltracer/src/main/java/io/opentracing/GlobalTracer.java]: There was some discussion about aligning on SLF4j. This will be the first use of logging in this project, so we might want to make a decision on that.

**Comment by wu-sheng** [comment]
+1 ` a separate module`

File: opentracing-globaltracer/src/main/java/io/opentracing/GlobalTracer.java (hunk 1 above)
Comment (wu-sheng) [inline opentracing-globaltracer/src/main/java/io/opentracing/GlobalTracer.java]: @sjoerdtalsma @bhs  Which log component do we suppose to add, it is a difficult choice, I think.

Comment (sjoerdtalsma) [inline opentracing-globaltracer/src/main/java/io/opentracing/GlobalTracer.java]: I think for libraries, the less transitive dependencies introduced the better.
I've been bitten too often by conflicting apache.commons.logging versions in the good old days!

I would only introduce a dependency to slf4j if there were measurable significant performance gains to back that up. I tried to be sparse with log statements and there shouldn't be many for the GlobalTracer.

Just my opinion of course, open to other arguments.

**Comment by sjoerdtalsma** [comment]
@bhs
> I would also propose that we choose a Java package name that can stick if we want to move this into the core module in the future.

Currently I put it in the `io.opentracing` java package as there's only a single `GlobalTracer` class.
I actually doubted about `io.opentracing.global` (similar to the current contrib sub-package).
Let's go with the majority vote on this one 😉 ..

File: opentracing-globaltracer/src/main/java/io/opentracing/GlobalTracer.java (hunk 1 above)
Comment (pavolloffay) [inline opentracing-globaltracer/src/main/java/io/opentracing/GlobalTracer.java]: +1 for JUL here.

**Comment by pavolloffay** [comment]
> Currently I put it in the io.opentracing java package as there's only a single GlobalTracer class.
I actually doubted about io.opentracing.global (similar to the current contrib sub-package).
Let's go with the majority vote on this one :wink:

What about `io.opentracing.util` as a place for globaltracer and maybe future spanmanager? But then we should probably introduce some "opentracing-util" artifact instead of `opentracing-globaltracer`.

**Comment by sjoerdtalsma** [comment]
I like the 'util' suggestion as it seems more stable to me.

**Comment by pavolloffay** [comment]
I did not follow the initial review of `globaltracer` therefore my following thoughts might not be relevant.

* It the update of a tracer considered as safe at runtime? Is `GlobalTracer` intended to be used for changing tracer impl at runtime or is it used only for the initial tracer acquiring? e.g. `GlobalTracer#register` can cause app crash. 

If that holds I think that tracer loading mechanism should be pluggable and `GlobalTracer#register` removed.

**Comment by sjoerdtalsma** [comment]
@pavolloffay I am afraid I don't quite understand your remark. If you have concerns about an app crash, please file [an issue](https://github.com/opentracing-contrib/java-globaltracer/issues) on the original repo, because that sounds dangerous.

Some background: GlobalTracer serves as a _holder_ for some 'global' configured tracer. By-default it will lazily try to resolve a Tracer from the ServiceLoader, but it is equally valid to `register` your own (e.g. if it requires explicit configuration that can not automatically be provided).
The GlobalTracer is a singleton because I was afraid people will create constants for it (similar to current logging patterns). Switching between incompatible tracers is not the intention here, but having predictable behaviour before, during and after application initialization is.

Again, if you feel the code is dangerous in any way, please file an issue so we can address it explicitly.

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+import java.util.logging.Logger;
+
+/**
+ * Forwards all methods to another tracer that can be configured in one of two ways:
+ * <ol>
+ * <li>Explicitly, calling {@link #register(Tracer)} with a configured tracer, or:</li>
+ * <li>Automatically using the Java {@link ServiceLoader} SPI mechanism to load a {@link Tracer} from the classpath.</li>
+ * </ol>
+ * <p>
+ * When the tracer is needed it is lazily looked up using the following rules:
```
Comment (bhs) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: I've been thinking about GlobalTracer... I definitely like the idea of having some sort of Tracer singleton in the core OT-Java repository / artifact (as it's essential to making something like #111 effective in practice), BUT I find this impl to be surprisingly subtle for an abstraction I'd expect to be simple.

My sense is that the complexity mostly stems from the requirement to lazy-init the GlobalTracer to avoid init-time race conditions. I get that the race conditions are to be avoided... but how about having the application programmer provide a signal about when to initialize GlobalTracer? Along with a convention that such initialization happens very early in the process lifetime, perhaps we could get rid of most/all of the lazy init code (as well as the reentrant lock, etc).

TL;DR I really want to keep it simple with GlobalTracer.

Comment (sjoerdtalsma) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: @bhs I am open to any simplification of course :wink: 
To clarify where I feel the complexity comes from; there are two opposing 'forces' driving this:
- It _really_ should be an actual singleton implementation as we won't be able to prevent people using globaltracer as constants.
- We cannot prevent references being obtained before the registration has happened.

I think the `ServiceLoader` auto-initialization could maybe be refactored out of the GlobalTracer itself though (we could create a separate issue for that either here after merging or in the original repo to be merged along).

Comment (pavolloffay) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: @sjoerdtalsma I will respond in this thread.

I also think that GlobalTracer should be simplified. 

> @pavolloffay I am afraid I don't quite understand your remark. If you have concerns about an app crash, please file an issue on the original repo, because that sounds dangerous.

The crash is not related to `GlobalTracer` itself. This tracer allows to re-register tracer at any time during the application lifetime which can cause app crash/exceptions. It allows you to register zipkin tracer and then hawkular tracer and if the code is using GlobalTracer like this 

```java
GlobalTracer.get()
    .buildSpan("foo")
    .asChildOf(child)
    .start();
```
 it would cause problems because child can be from a different tracer (and spec does not deal with this corner case).

Maybe `GlobalTracer.register()` should be only one-time  operation.

Comment (sjoerdtalsma) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: I understand your point and am open to simplification, but how would you propose to handle adding a wrapper tracer?
https://github.com/opentracing-contrib/java-globaltracer/issues/7

Comment (objectiser) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: We could simplify the requirement - add a new interface `WrapperTracer` and allow lazy instantiation of only one wrapper and one concrete Tracer - if wrapper is available it is then supplied the concrete tracer before end of initialisation phase.

And `register` is a once only operation (as Pavol suggested).

Comment (pavolloffay) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: What if somebody needs multiple wrappers/tee adapters? 

Would this be sufficient?
1. make `tracers = GlobalTracer.loadFromSpi()` public utility function which would return all possible tracers.
2. `GlobalTracer.register(tracer)` would throw checked exception if there is something alredy registered
3. `GlobalTracer.get()` reutrns null/noop? or registered tracer

this makes an application always responsible for tracer registration. @objectiser would it be sufficient for agent?

Comment (objectiser) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: Reason for suggesting single wrapper, is that I originally suggested a more complex approach supporting multiple wrapper tracers https://github.com/opentracing-contrib/java-globaltracer/issues/7#issuecomment-286706102 - but I think the complexity lead to the discussion about using an `update` method, which leads to the problem of tracers being updated after first used, which is not desirable.

I still like the lazy instantiation approach as it gives time for the app to explicitly register its own tracer before first use.

Comment (pavolloffay) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: >  which leads to the problem of tracers being updated after first used, which is not desirable.

Because of this I proposing to change `GlobalTracer` to one-time action and make the app responsible for initializing it.

Comment (objectiser) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: Problem is that there may be no central part of the app to make this decision - e.g. a service instrumented with multiple framework integrations - which part takes control of initialising the global tracer? Each framework integration is simply a consumer of the global tracer.

Comment (sjoerdtalsma) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: ^-- this.
This responsibility lies with the final application and putting up too many restrictions because they may abuse the responsibility doesn't feel right to me (unless we learn that abuse becomes problematic).

Comment (sjoerdtalsma) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: > GlobalTracer.get() reutrns null/noop? or registered tracer

What if there's a class that does 
```java
private static final Tracer GLOBAL = GlobalTracer.get();
```

And the class gets loaded after initialization, everything is okay, the code is moved into a library (or gets forgotten about) and somehow the class gets loaded _before_ the tracer initialization?
As tracing is so similar to logging for many people, I feel this use-case cannot be ignored.

Comment (sjoerdtalsma) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: I'm getting a bit of a _deja vu_ by the way :wink: which means these probably are valid concerns.

Comment (sjoerdtalsma) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: I suggest creating an issue in [opentracing-contrib/java-globaltracer](https://github.com/opentracing-contrib/java-globaltracer/issues) to mitigate
>the problem of tracers being updated after first used, which is not desirable

then we can discuss alternatives solutions there (e.g. we could also simply flag GT as 'used' upon use).
I am willing to postpone merging this until said issue is resolved.

Comment (objectiser) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: I can create a separate issue for this, but wondering whether this migration could occur based on the version before https://github.com/opentracing-contrib/java-globaltracer/issues/7 - and then continue this discussion in that PR?
Not sure it is worth holding up the move in general.

Comment (sjoerdtalsma) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: @objectiser The corresponding PR https://github.com/opentracing-contrib/java-globaltracer/pull/8 has not been merged yet, so we're free to choose which issue we merge first.

Comment (objectiser) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: @sjoerdtalsma Sorry wasn't clear - this PR is based on the unmerged work that introduces the `update` function etc. Wondering whether we should just migrate the latest committed version of the global tracer and then work of that as a base for further discussion?

Comment (pavolloffay) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: > Problem is that there may be no central part of the app to make this decision - e.g. a service instrumented with multiple framework integrations - which part takes control of initialising the global tracer? Each framework integration is simply a consumer of the global tracer.

My assumption is that app startup is controlled and therefore GlobalTracer initialization too. Problem would be if framework instrumentation (which uses GlobalTracer) would wrap the tracer and wanted to propagate the change to different part of the system. 
@objectiser is there any specific instrumentation which does so?

**Comment by tedsuo** [comment]
So happy to see this here!

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java (hunk 2 above)
Comment (bhs) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: My two cents is that GlobalTracer should aspire to be nothing more than a global Tracer. All of the concerns about wrapping, etc, can be handled by layers above the core OT API... the only reason I want GlobalTracer to be part of OT proper is that we do want a global/singleton Tracer *accessor* if we expect library instrumentation code to be easy to write and maintain.

I am a little more sympathetic about the ServiceLoader functionality (than the wrapper), but still feel like it doesn't *need* to be here to satisfy the core requirement of library instrumentation, and thus can be left out for now. Adding it in later will increase the line count but shouldn't change the API to access the global Tracer.

I also think it is fine to make the GlobalTracer "write-once"... i.e., no `update(...)` functionality. As was mentioned previously, even if we support that in the GlobalTracer, it could wreak havoc with Tracers/Spans that assume they know the types of other Spans and SpanContexts propagating through the process at the time.

PS: I'm sorry to take so long to follow up on this – my calendar is insane right now and I'm falling behind on ~everything.

Comment (objectiser) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: I'm happy with a "write-once" approach, so the app (when in control of tracer initialisation) would register the (potentially wrapped) tracer.

However I would also like to allow for more complex implicit tracer initialisation. If we have a `TracerProvider` (or `TracerResolver`) API - if no `Tracer` has been explicitly registered then first attempt to service load a `TracerProvider` (which may implement its own strategy for creating the `Tracer`) and finally if not found, then use a default `TracerProvider` which uses the `ServiceLoader` to retrieve a `Tracer`?

So by default we still have the simple case of explicit vs implicit service load of Tracer, but allows flexibility for a more complex strategy to be used.

Comment (yurishkuro) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: @objectiser if the GlobalTracer API in OT-proper was simply SetOnce() and Get(), could the implicit loading that you want be just another module (e.g. from contrib) that people who need it would include in their app?

Comment (objectiser) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: Yes that would work fine aswell.

Comment (bhs) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: ^^^ I think this would be a wise place to start. Far, far easier to add complexity than remove it. :)

To be clear, (most of) the more sophisticated features we're discussing here *do* sound valuable... I just want to keep them out of the core OT-java repo unless there is no alternative.

Comment (sjoerdtalsma) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: I'll see if I can persuade my wife for some cycles this evening to update the PR :wink:

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            LOGGER.log(Level.FINE, "Attempted to register the GlobalTracer as delegate of itself.");
+            return; // no-op
+        } else if (globalTracer != null && !globalTracer.equals(tracer)) {
+            throw new IllegalStateException("There is already a current globalTracer registered.");
+        }
+        globalTracer = tracer;
+    }
+
+    private static Tracer globalOrNoopTracer() {
+        return globalTracer == null ? NoopTracerFactory.create() : globalTracer;
```
Comment (bhs) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: why do we re-create() the NoopTracer every time this is called? seems to me that we could do that work once at static init time.

Comment (sjoerdtalsma) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: I think that's more a comment about the design of the NoopTracer (and its Factory) in Java.
Although the method is called `create` it merely returns the constant.

**Comment by sjoerdtalsma** [comment]
@bhs 
> i.e., returning the Tracer instance that was registered() at get()-time.

Two points:
1. That would not work with libraries that get loaded before the global tracer is registered and work with a constant reference.
2. The delegation forces using the `Tracer` interface.

File: opentracing-util/pom.xml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-all</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>org.hamcrest</groupId>
```
Comment (pavolloffay) [inline opentracing-util/pom.xml]: Is it necessary to introduce this for those simple asserts?

File: opentracing-util/src/test/java/io/opentracing/util/GlobalTracerTest.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        } catch (RuntimeException expected) {
+            assertThat("Duplicate registration message", expected.getMessage(), is(notNullValue()));
+        }
+    }
+
+    /**
+     * Check leniency for duplicate registration with the same tracer by mistake.
+     */
+    @Test
+    public void testMultipleRegistrations_sameTracer() {
```
Comment (pavolloffay) [inline opentracing-util/src/test/java/io/opentracing/util/GlobalTracerTest.java]: This actually does not test anything, assert on `GlobalTracer.get()`?

**Comment by pavolloffay** [comment]
Why it cannot be simplified to something like: https://pastebin.com/jijhG1NE

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+import io.opentracing.propagation.Format;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Forwards all methods to another tracer that can be configured by calling {@link #register(Tracer)}.
+ * <p>
+ * When the tracer is needed it delegates all tracing methods to
+ * either the <em>registered</em> global tracer or the {@link NoopTracer}.
```
Comment (yurishkuro) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: this sentence is mostly redundant. I would just say that "if register() is never called, the default NoopTracer is used".

Comment (yurishkuro) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: I would also like to see the comments mention the set-once semantics, and explain the limitations.

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+     * all behave as if obtained <em>after</em> initialization once properly initialized.<br>
+     * As a minor additional benefit it makes it harder to circumvent the {@link Tracer} API.
+     */
+    private static final GlobalTracer INSTANCE = new GlobalTracer();
+
+    /**
+     * The resolved {@link Tracer} to delegate to or the {@link NoopTracer} if none was registered yet.
+     * Never {@code null}.
+     */
+    private static volatile Tracer globalTracer = NoopTracerFactory.create();
```
Comment (yurishkuro) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: do we need volatile if we're agreeing on set-once semantic?

File: opentracing-util/src/test/java/io/opentracing/util/GlobalTracerTest.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            globalTracerField.setAccessible(true);
+            globalTracerField.set(null, tracer);
+            globalTracerField.setAccessible(false);
+        } catch (Exception e) {
+            throw new RuntimeException("Error reflecting globalTracer: " + e.getMessage(), e);
+        }
+    }
+
+    @Before
+    @After
```
Comment (sjoerdtalsma) [inline opentracing-util/src/test/java/io/opentracing/util/GlobalTracerTest.java]: I also like to leave the global tracer unregistered _after_ the test, that's all 😉 .

File: opentracing-util/src/test/java/io/opentracing/util/GlobalTracerTest.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    @Test
+    public void testGet_SingletonReference() {
+        Tracer tracer = GlobalTracer.get();
+        assertThat(tracer, is(instanceOf(GlobalTracer.class)));
+        assertThat(tracer, is(sameInstance(GlobalTracer.get())));
+    }
+
+    @Test
```
Comment (sjoerdtalsma) [inline opentracing-util/src/test/java/io/opentracing/util/GlobalTracerTest.java]: The first register should succeed, the second shouldn't.
Expecting the runtime exception is therefore _weaker_ than the test for the exception to occur in the second call. Also I get to test for a non-null exception message, although I agree there's no high value here.

File: opentracing-util/src/test/java/io/opentracing/util/GlobalTracerTest.java (hunk 5 above)
Comment (sjoerdtalsma) [inline opentracing-util/src/test/java/io/opentracing/util/GlobalTracerTest.java]: It 'tests' that registering the same tracer twice is allowed.

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java (hunk 7 above)
Comment (sjoerdtalsma) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: I forget in which Java version setting an object reference became thread-safe.
Therefore I usually make concurrent read/writes for a reference `volatile` but I'd love to hear a definitive answer to this question

File: opentracing-util/pom.xml (hunk 4 above)
Comment (sjoerdtalsma) [inline opentracing-util/pom.xml]: I guess not. I'm just more used to it. I'll see if it can be dropped without too much hassle.

**Comment by pavolloffay** [comment]
@sjoerdtalsma thanks for the explanation :+1: Then it makes sense. <sub>I think I missed my coffee.</sub>

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java (hunk 7 above)
Comment (yurishkuro) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: my point is if we say upfront that this must be set once during application startup, then we do not need to worry about threads at all.

Comment (sjoerdtalsma) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: To avoid doubt I created a unit test which failed (allowed multiple register() calls to make it through).
The simple solution was to make `register()` synchronized and remove `volatile` as indeed no longer necessary.

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+     * Registration is a one-time operation, attempting to call it more often will result in a runtime exception.
+     * <p>
+     * Every application intending to use the global tracer is responsible for registering it once
+     * during its initialization.
+     *
+     * @param tracer Tracer to use as global tracer.
+     * @throws RuntimeException if there is already a current tracer registered
+     */
+    public static synchronized void register(final Tracer tracer) {
+        if (tracer == null) throw new NullPointerException("Cannot register GlobalTracer <null>.");
```
Comment (yurishkuro) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: nit: enclose throw in `{}`

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        else if (tracer instanceof GlobalTracer) {
```
Comment (yurishkuro) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: don't need `else`

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+            LOGGER.log(Level.FINE, "Attempted to register the GlobalTracer as delegate of itself.");
+            return; // no-op
+        } else if (!(globalTracer instanceof NoopTracer) && !globalTracer.equals(tracer)) {
```
Comment (yurishkuro) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: don't need `else`

**Comment by yurishkuro** [comment]
lgtm

**Comment by objectiser** [comment]
@bhs Any objections to merging this one soon and doing a release?

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+     * all behave as if obtained <em>after</em> initialization once properly initialized.<br>
+     * As a minor additional benefit it makes it harder to circumvent the {@link Tracer} API.
+     */
+    private static final GlobalTracer INSTANCE = new GlobalTracer();
+
+    /**
+     * The registered {@link Tracer} delegate or the {@link NoopTracer} if none was registered yet.
+     * Never {@code null}.
+     */
+    private static Tracer globalTracer = NoopTracerFactory.create();
```
Comment (yurishkuro) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: nit: this would probably be better named just `tracer`

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    @Override
+    public <C> SpanContext extract(Format<C> format, C carrier) {
+        return globalTracer.extract(format, carrier);
+    }
+
+    @Override
+    public String toString() {
+        return GlobalTracer.class.getSimpleName();
```
Comment (yurishkuro) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: should this be `globalTracer` instead of `GlobalTracer`? Or perhaps both, `GlobalTracer(com....MyTracer)`

Comment (objectiser) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: +1 both - good to know using global tracer, but also what tracer it is wrapping.

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java (hunk 13 above)
Comment (sjoerdtalsma) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: Good idea, it's been renamed a number of times already so one more won't hurt 😉

File: opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+     */
+    public static synchronized void register(final Tracer tracer) {
+        if (tracer == null) {
+            throw new NullPointerException("Cannot register GlobalTracer <null>.");
+        }
+        if (tracer instanceof GlobalTracer) {
+            LOGGER.log(Level.FINE, "Attempted to register the GlobalTracer as delegate of itself.");
+            return; // no-op
+        }
+        if (GlobalTracer.tracer instanceof NoopTracer) {
```
Comment (bhs) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: sorry if I'm reading this wong... but where is the code path where we actually register a non-NoopTracer?

Comment (yurishkuro) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: https://github.com/opentracing/opentracing-java/pull/109/files#diff-46890bf250007136c56d965126d8566eR51

```java
private static Tracer tracer = NoopTracerFactory.create();
```

Comment (bhs) [inline opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java]: sorry, I was (too quickly) reading this line as `tracer instanceof NoopTracer`, not `GlobalTracer.tracer instanceof NoopTracer`.

LG.

**Comment by bhs** [comment]
@sjoerdtalsma if there's urgency, we could probably merge ~now / today... otherwise I'd wait 24h since this is a more substantial change than most.

**Comment by sjoerdtalsma** [comment]
@bhs I'm no collaborator on the `opentracing-java` git repo, so I can't merge this in myself.

**Comment by bhs** [comment]
Done – thanks so much, @sjoerdtalsma!

**Comment by bhs** [comment]
@sjoerdtalsma oh, one other thing: want to make a small change to the OT-java README that mentions the GlobalTracer?

**Comment by objectiser** [comment]
@sjoerdtalsma Good job!

@bhs Time for a release :) - would be good to start updating the framework integrations and agent.

**Comment by bhs** [comment]
@objectiser I just pushed a release tag... we should be good to go soon with `0.21.0`.

**Comment by wu-sheng** [comment]
👏 @sjoerdtalsma Glad to have this in `opentracing-java`, finally.

**Comment by sjoerdtalsma** [comment]
Thanks for the reviews everybody!
